ouroboros-consensus/test-consensus/Test/ThreadNet/RealPBFT.hs

```diff
@@ -669,45 +1023,54 @@ shrinkTestConfigSlotsOnly TestConfig
 genNodeRekeys
   :: PBftParams
   -> NodeJoinPlan
+  -> NodeTopology
   -> NumSlots
   -> NodeRestarts
   -> Gen NodeRestarts
-genNodeRekeys params (NodeJoinPlan njp) numSlots@(NumSlots t)
+genNodeRekeys params nodeJoinPlan nodeTopology numSlots@(NumSlots t)
   nodeRestarts@(NodeRestarts nrs)
   | t <= 0    = pure nodeRestarts
   | otherwise =
-    -- The necessary conditions are pretty rare, so favor adding a 'NodeRekey'
-    -- when we can. But not always.
+    -- The necessary conditions are relatively rare, so favor adding a
+    -- 'NodeRekey' when we can. But not always.
     (\x -> frequency [(2, pure nodeRestarts), (8, x)]) $
     -- TODO rekey nodes other than the last
     -- TODO rekey more than one node
     -- TODO rekey a node in a slot other than its join slot
     case Map.lookupMax njp of
-      Just (nid, jslot)
+      Just (cid, jslot)
             -- last node joins after first epoch, ...
           | jslot >= beginSecondEpoch
             -- ... and could instead join unproblematically at the latest time
-            -- the delegation certificate would mature
+            -- the delegation certificate would mature ...
           , latestPossibleDlgMaturation pbftSecurityParam numCoreNodes jslot
-              <= lastSlot
+              < sentinel
           , let nodeJoinPlan' =
-                  NodeJoinPlan $ Map.insert nid (jslot + twoK) njp
-          , Ref.viable params lastSlot nodeJoinPlan' Ref.emptyState
-          , Ref.deterministicPlan numSlots nodeJoinPlan'
+                  NodeJoinPlan $ Map.insert cid (jslot + twoK) njp
+          , Ref.definitelyEnoughBlocks params $
+            Ref.simulate params nodeJoinPlan' numSlots
+            -- ... and does not join in the same slot as the leader of the next
+            -- slot unless they are neighbors (otherwise the dlg cert tx might
+            -- not reach it soon enough)
+          , let nextLeader = Ref.mkLeaderOf params $ succ jslot
+          , jslot /= coreNodeIdJoinSlot nodeJoinPlan nextLeader ||
+            cid `elem` coreNodeIdNeighbors nodeTopology nextLeader
           -> pure $ NodeRestarts $
```

Comment:
We also need to prevent the rekeying node from restarting too soon, since it might need to propagate its dlg cert tx more than once.